Proxy to HistoricalRPCService

alecps opened the pull request:
- Minor edits to proxy out to the HistoricalRPCService based on cel2 time rather than bedrock block. 
- Adds support for TraceCall proxying to HistoricalRPCService

Fixes https://github.com/celo-org/celo-blockchain-planning/issues/386

Testing:
- Manually tested all rpc endpoints (see doc here https://www.notion.so/clabsco/HistoricalRPCService-Testing-Scratchpad-10aea6297b89809789bde885503a6210?showMoveTo=true&saveParent=true) 
- Working on adding unit tests, but have gotten stuck. See https://github.com/celo-org/op-geth/pull/238

Commit:
add isPreCel2 function and use instead of isOptimismPreBedrock

Commit:
add support for traceCall

carterqw2 commented on eth/tracers/api.go:
```diff
@@ -961,8 +961,16 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
 
-	if api.backend.ChainConfig().IsOptimismPreBedrock(block.Number()) {
-		return nil, errors.New("l2geth does not have a debug_traceCall method")
+	if api.backend.ChainConfig().IsPreCel2(block.Time()) {
```
Perhaps we can combine both conditionals.

karlb replied:
I don't see a simple way to do it since we have three different results:
* Return result from local node
* Return proxies result
* Return `rpc.ErrNoHistoricalFallback`

piersy replied:
If you combined them then you'd have to repeat one of the conditions though, right?

piersy commented on params/config.go:
```diff
@@ -769,6 +769,10 @@ func (c *ChainConfig) IsCel2(time uint64) bool {
 }
 
+func (c *ChainConfig) IsPreCel2(time uint64) bool {
+	return !c.IsCel2(time)
```
This doesn't look right to me, IsPreCel2 will return true in situations where the cel2 fork was not set

karlb replied:
Yes, Optimism has the `IsOptimism` check in addition. For us, it should probably be something like `c.Cel2Time != nil && !c.IsCel2(time)`

piersy replied:
I'm just testing this now

karlb reviewed (changes requested):
Do we already understand why we get the "implement me" test failure?

piersy commented:
> Do we already understand why we get the "implement me" test failure?

I don't, and I'm looking into that now

karlb committed:
Fix IsPreCel2 condition for non-Celo case

If c.Cel2Time is not set, we IsPreCel2 should always be false.

karlb commented:
Fixing the IsPreCel2 condition fixes the two failing tests locally. Let's see if CI agrees.

piersy commented:
> Fixing the IsPreCel2 condition fixes the two failing tests locally. Let's see if CI agrees.

No all the pre-existing historical rpc tests are failing because they don't have cel2 set in their test config. But setting cel2 is causing further problems (at least in `TestEthClientHistoricalBackend`)

gastonponti commented on params/config.go:
```diff
@@ -769,6 +769,10 @@ func (c *ChainConfig) IsCel2(time uint64) bool {
 }
 
+func (c *ChainConfig) IsPreCel2(time uint64) bool {
+	return c.Cel2Time != nil && !c.IsCel2(time)
```
We have a config.Celo now due to the baseFeeFloor, if we want to mimic the same logic that the optimism's team is doing.

karlb committed:
Make historical tests pass with Cel2 setup

The original tests worked on `IsOptimismPreBedrock` and need some
adaptation to work in the `IsPreCel2` case.

Setting cel2Time is necessary for that to work. Unfortunately, setting
that also triggers another condition in our codebase that we don't
actually want for the test. We skip setting GasLimit and Difficulty to
stay compatible with how genesis blocks have been created on Celo chains
in the past. It would be preferable to detect the necessity for that in
a different way, but I'll leave that for a future PR.

karlb commented:
I forced-pushed to rebase on https://github.com/celo-org/op-geth/pull/239, which helped getting the tests to pass. The only actually new commit in the recent push is https://github.com/celo-org/op-geth/pull/236/commits/cd9a87fc64fda450a20b7e039669943d84ce2277. With that the tests should pass and this PR should be ready.

karlb reviewed (approved):
I'll approve to undo my "request changes", but someone else should review before merging, because I did the last changes myself.
